File: src/main/java/eu/deyanix/pi4juart/controller/SerialController.java
package eu.deyanix.pi4juart.controller;

import com.pi4j.io.serial.Serial;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.io.IOException;

@RestController
public class SerialController {
	private final Serial serial;

	public SerialController(Serial serial) {
		this.serial = serial;
	}

	@PostMapping("/serial")
	public void write(@RequestBody String text) throws IOException {
		serial.getOutputStream().write(text.getBytes());
		serial.getOutputStream().write('\n');
		serial.getOutputStream().write('\r');
	}
}

File: src/main/kotlin/eu/deyanix/pi4juart/controller/SerialController.kt
package eu.deyanix.pi4juart.controller

import eu.deyanix.pi4juart.service.SerialReader
import eu.deyanix.pi4juart.service.SerialService
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.web.bind.annotation.*
import java.io.IOException

@Tag(name = "Serial")
@RestController
class SerialController(
	private val serialService: SerialService,
	private val serialReader: SerialReader,
) {
	@GetMapping("/serial/read")
	@Throws(IOException::class)
	fun read(): String {
		return serialReader.readLine()
	}

	@GetMapping("/serial/lines")
	@Throws(IOException::class)
	fun nextLines(): Array<String> {
		return serialReader.nextLines()
	}

	@PostMapping("/serial")
	@Throws(IOException::class)
	fun write(@RequestBody text: String) {
		serialService.println(text)
	}
}
